packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1417,6 +1427,39 @@ abstract class ResidentRunner {
     appFinished();
   }
 
+  bool _reportedDebuggers = false;
+
+  void printDebuggerList({ bool includeObservatory = true, bool includeDevtools = true }) {
+    final DevToolsServerAddress devToolsServerAddress = activeDevToolsServer();
+    if (devToolsServerAddress == null) {
+      includeDevtools = false;
+    }
+    for (final FlutterDevice device in flutterDevices) {
+      if (device.vmService == null) {
+        continue;
+      }
+      if (includeObservatory) {
+        // Caution: This log line is parsed by device lab tests.
+        globals.printStatus(
+          'An Observatory debugger and profiler on ${device.device.name} is available at: '
+          '${device.vmService.httpAddress}',
+        );
+      }
+      if (includeDevtools) {
+        final Uri uri = devToolsServerAddress.uri?.replace(
+          queryParameters: <String, dynamic>{'uri': '${device.vmService.httpAddress}'},
```

Comment:
will httpAddress work for flutter web apps? I think we'd need to use the wsAddress for those

Reply:
Dunno, I didn't change this. I just used factored out the common code in run_hot and run_cold into this method, the logic is basically identical to what it was before (as far as I can tell... maybe I made a typo?).